#### Load the files

In [248]:
import numpy as np
import pandas as pd
import glob 
from datetime import datetime

In [249]:


files = glob.glob('../data/Fitabase Data 4.12.16-5.12.16/*')
dailyActivity_merged = pd.read_csv(files[0])
dailyCalories_merged = pd.read_csv(files[1])
dailyIntensities_merged = pd.read_csv(files[2])
dailySteps_merged = pd.read_csv(files[3])
heartrate_seconds_merged = pd.read_csv(files[4])
hourlyCalories_merged = pd.read_csv(files[5])
hourlyIntensities_merged = pd.read_csv(files[6])
hourlySteps_merged = pd.read_csv(files[7])
minuteCaloriesNarrow_merged = pd.read_csv(files[8])
minuteCaloriesWide_merged = pd.read_csv(files[9])
minuteIntensitiesNarrow_merged = pd.read_csv(files[10])
minuteIntensitiesWide_merged = pd.read_csv(files[11])
minuteMETsNarrow_merged = pd.read_csv(files[12])
minuteSleep_merged = pd.read_csv(files[13])
minuteStepsNarrow_merged = pd.read_csv(files[14])
minuteStepsWide_merged = pd.read_csv(files[15])
sleepDay_merged = pd.read_csv(files[16])
weightLogInfo_merged = pd.read_csv(files[17])


#### Get a quick summary of structure and content of each files

In [250]:
# Create a function to help show relevant information
def pstr(df):
    # return df.info(null_counts=True, verbose=True), df.apply(lambda x: [x.unique()]), df.describe()
    print(df.info(null_counts=True, verbose=True))
    display(df.apply(lambda x: [x.unique()[:5]]).T)
    display(df.describe())

In [251]:
pstr(dailyActivity_merged)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 940 entries, 0 to 939
Data columns (total 15 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Id                        940 non-null    int64  
 1   ActivityDate              940 non-null    object 
 2   TotalSteps                940 non-null    int64  
 3   TotalDistance             940 non-null    float64
 4   TrackerDistance           940 non-null    float64
 5   LoggedActivitiesDistance  940 non-null    float64
 6   VeryActiveDistance        940 non-null    float64
 7   ModeratelyActiveDistance  940 non-null    float64
 8   LightActiveDistance       940 non-null    float64
 9   SedentaryActiveDistance   940 non-null    float64
 10  VeryActiveMinutes         940 non-null    int64  
 11  FairlyActiveMinutes       940 non-null    int64  
 12  LightlyActiveMinutes      940 non-null    int64  
 13  SedentaryMinutes          940 non-null    int64  
 14  Calories  

C:\Users\User\AppData\Local\Temp\ipykernel_10580\3310633354.py:4: FutureWarning: null_counts is deprecated. Use show_counts instead
  print(df.info(null_counts=True, verbose=True))


0
Id                        [1503960366, 1624580081, 1644430081, 184450507...
ActivityDate              [4/12/2016, 4/13/2016, 4/14/2016, 4/15/2016, 4...
TotalSteps                               [13162, 10735, 10460, 9762, 12669]
TotalDistance             [8.5, 6.96999979019165, 6.73999977111816, 6.28...
TrackerDistance           [8.5, 6.96999979019165, 6.73999977111816, 6.28...
LoggedActivitiesDistance  [0.0, 1.95959603786469, 4.08169221878052, 2.78...
VeryActiveDistance        [1.87999999523163, 1.57000005245209, 2.4400000...
ModeratelyActiveDistance  [0.550000011920929, 0.689999997615814, 0.40000...
LightActiveDistance       [6.05999994277954, 4.71000003814697, 3.9100000...
SedentaryActiveDistance   [0.0, 0.0099999997764825, 0.0199999995529652, ...
VeryActiveMinutes                                      [25, 21, 30, 29, 36]
FairlyActiveMinutes                                    [13, 19, 11, 34, 10]
LightlyActiveMinutes                              [328, 217, 181, 209, 221]
SedentaryMinutes                                 [728, 776, 1218, 726, 773]
Calories                                     [1985, 1797, 1776, 1745, 1863]

Id    TotalSteps  TotalDistance  TrackerDistance  \
count  9.400000e+02    940.000000     940.000000       940.000000   
mean   4.855407e+09   7637.910638       5.489702         5.475351   
std    2.424805e+09   5087.150742       3.924606         3.907276   
min    1.503960e+09      0.000000       0.000000         0.000000   
25%    2.320127e+09   3789.750000       2.620000         2.620000   
50%    4.445115e+09   7405.500000       5.245000         5.245000   
75%    6.962181e+09  10727.000000       7.712500         7.710000   
max    8.877689e+09  36019.000000      28.030001        28.030001   

       LoggedActivitiesDistance  VeryActiveDistance  ModeratelyActiveDistance  \
count                940.000000          940.000000                940.000000   
mean                   0.108171            1.502681                  0.567543   
std                    0.619897            2.658941                  0.883580   
min                    0.000000            0.000000                  0.000000   
25%                    0.000000            0.000000                  0.000000   
50%                    0.000000            0.210000                  0.240000   
75%                    0.000000            2.052500                  0.800000   
max                    4.942142           21.920000                  6.480000   

       LightActiveDistance  SedentaryActiveDistance  VeryActiveMinutes  \
count           940.000000               940.000000         940.000000   
mean              3.340819                 0.001606          21.164894   
std               2.040655                 0.007346          32.844803   
min               0.000000                 0.000000           0.000000   
25%               1.945000                 0.000000           0.000000   
50%               3.365000                 0.000000           4.000000   
75%               4.782500                 0.000000          32.000000   
max              10.710000                 0.110000         210.000000   

       FairlyActiveMinutes  LightlyActiveMinutes  SedentaryMinutes  \
count           940.000000            940.000000        940.000000   
mean             13.564894            192.812766        991.210638   
std              19.987404            109.174700        301.267437   
min               0.000000              0.000000          0.000000   
25%               0.000000            127.000000        729.750000   
50%               6.000000            199.000000       1057.500000   
75%              19.000000            264.000000       1229.500000   
max             143.000000            518.000000       1440.000000   

          Calories  
count   940.000000  
mean   2303.609574  
std     718.166862  
min       0.000000  
25%    1828.500000  
50%    2134.000000  
75%    2793.250000  
max    4900.000000

In [252]:
pstr(dailyCalories_merged)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 940 entries, 0 to 939
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Id           940 non-null    int64 
 1   ActivityDay  940 non-null    object
 2   Calories     940 non-null    int64 
dtypes: int64(2), object(1)
memory usage: 22.2+ KB
None


C:\Users\User\AppData\Local\Temp\ipykernel_10580\3310633354.py:4: FutureWarning: null_counts is deprecated. Use show_counts instead
  print(df.info(null_counts=True, verbose=True))


0
Id           [1503960366, 1624580081, 1644430081, 184450507...
ActivityDay  [4/12/2016, 4/13/2016, 4/14/2016, 4/15/2016, 4...
Calories                        [1985, 1797, 1776, 1745, 1863]

Id     Calories
count  9.400000e+02   940.000000
mean   4.855407e+09  2303.609574
std    2.424805e+09   718.166862
min    1.503960e+09     0.000000
25%    2.320127e+09  1828.500000
50%    4.445115e+09  2134.000000
75%    6.962181e+09  2793.250000
max    8.877689e+09  4900.000000

In [253]:
pstr(dailyIntensities_merged)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 940 entries, 0 to 939
Data columns (total 10 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Id                        940 non-null    int64  
 1   ActivityDay               940 non-null    object 
 2   SedentaryMinutes          940 non-null    int64  
 3   LightlyActiveMinutes      940 non-null    int64  
 4   FairlyActiveMinutes       940 non-null    int64  
 5   VeryActiveMinutes         940 non-null    int64  
 6   SedentaryActiveDistance   940 non-null    float64
 7   LightActiveDistance       940 non-null    float64
 8   ModeratelyActiveDistance  940 non-null    float64
 9   VeryActiveDistance        940 non-null    float64
dtypes: float64(4), int64(5), object(1)
memory usage: 73.6+ KB
None


C:\Users\User\AppData\Local\Temp\ipykernel_10580\3310633354.py:4: FutureWarning: null_counts is deprecated. Use show_counts instead
  print(df.info(null_counts=True, verbose=True))


0
Id                        [1503960366, 1624580081, 1644430081, 184450507...
ActivityDay               [4/12/2016, 4/13/2016, 4/14/2016, 4/15/2016, 4...
SedentaryMinutes                                 [728, 776, 1218, 726, 773]
LightlyActiveMinutes                              [328, 217, 181, 209, 221]
FairlyActiveMinutes                                    [13, 19, 11, 34, 10]
VeryActiveMinutes                                      [25, 21, 30, 29, 36]
SedentaryActiveDistance   [0.0, 0.0099999997764825, 0.0199999995529652, ...
LightActiveDistance       [6.05999994277954, 4.71000003814697, 3.9100000...
ModeratelyActiveDistance  [0.550000011920929, 0.689999997615814, 0.40000...
VeryActiveDistance        [1.87999999523163, 1.57000005245209, 2.4400000...

Id  SedentaryMinutes  LightlyActiveMinutes  \
count  9.400000e+02        940.000000            940.000000   
mean   4.855407e+09        991.210638            192.812766   
std    2.424805e+09        301.267437            109.174700   
min    1.503960e+09          0.000000              0.000000   
25%    2.320127e+09        729.750000            127.000000   
50%    4.445115e+09       1057.500000            199.000000   
75%    6.962181e+09       1229.500000            264.000000   
max    8.877689e+09       1440.000000            518.000000   

       FairlyActiveMinutes  VeryActiveMinutes  SedentaryActiveDistance  \
count           940.000000         940.000000               940.000000   
mean             13.564894          21.164894                 0.001606   
std              19.987404          32.844803                 0.007346   
min               0.000000           0.000000                 0.000000   
25%               0.000000           0.000000                 0.000000   
50%               6.000000           4.000000                 0.000000   
75%              19.000000          32.000000                 0.000000   
max             143.000000         210.000000                 0.110000   

       LightActiveDistance  ModeratelyActiveDistance  VeryActiveDistance  
count           940.000000                940.000000          940.000000  
mean              3.340819                  0.567543            1.502681  
std               2.040655                  0.883580            2.658941  
min               0.000000                  0.000000            0.000000  
25%               1.945000                  0.000000            0.000000  
50%               3.365000                  0.240000            0.210000  
75%               4.782500                  0.800000            2.052500  
max              10.710000                  6.480000           21.920000

In [254]:
pstr(dailySteps_merged)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 940 entries, 0 to 939
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Id           940 non-null    int64 
 1   ActivityDay  940 non-null    object
 2   StepTotal    940 non-null    int64 
dtypes: int64(2), object(1)
memory usage: 22.2+ KB
None


C:\Users\User\AppData\Local\Temp\ipykernel_10580\3310633354.py:4: FutureWarning: null_counts is deprecated. Use show_counts instead
  print(df.info(null_counts=True, verbose=True))


0
Id           [1503960366, 1624580081, 1644430081, 184450507...
ActivityDay  [4/12/2016, 4/13/2016, 4/14/2016, 4/15/2016, 4...
StepTotal                   [13162, 10735, 10460, 9762, 12669]

Id     StepTotal
count  9.400000e+02    940.000000
mean   4.855407e+09   7637.910638
std    2.424805e+09   5087.150742
min    1.503960e+09      0.000000
25%    2.320127e+09   3789.750000
50%    4.445115e+09   7405.500000
75%    6.962181e+09  10727.000000
max    8.877689e+09  36019.000000

In [255]:
pstr(heartrate_seconds_merged)

C:\Users\User\AppData\Local\Temp\ipykernel_10580\3310633354.py:4: FutureWarning: null_counts is deprecated. Use show_counts instead
  print(df.info(null_counts=True, verbose=True))


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2483658 entries, 0 to 2483657
Data columns (total 3 columns):
 #   Column  Non-Null Count    Dtype 
---  ------  --------------    ----- 
 0   Id      2483658 non-null  int64 
 1   Time    2483658 non-null  object
 2   Value   2483658 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 56.8+ MB
None


0
Id     [2022484408, 2026352035, 2347167796, 402033265...
Time   [4/12/2016 7:21:00 AM, 4/12/2016 7:21:05 AM, 4...
Value                           [97, 102, 105, 103, 101]

Id         Value
count  2.483658e+06  2.483658e+06
mean   5.513765e+09  7.732842e+01
std    1.950224e+09  1.940450e+01
min    2.022484e+09  3.600000e+01
25%    4.388162e+09  6.300000e+01
50%    5.553957e+09  7.300000e+01
75%    6.962181e+09  8.800000e+01
max    8.877689e+09  2.030000e+02

In [256]:
pstr(hourlyCalories_merged)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22099 entries, 0 to 22098
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Id            22099 non-null  int64 
 1   ActivityHour  22099 non-null  object
 2   Calories      22099 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 518.1+ KB
None


C:\Users\User\AppData\Local\Temp\ipykernel_10580\3310633354.py:4: FutureWarning: null_counts is deprecated. Use show_counts instead
  print(df.info(null_counts=True, verbose=True))


0
Id            [1503960366, 1624580081, 1644430081, 184450507...
ActivityHour  [4/12/2016 12:00:00 AM, 4/12/2016 1:00:00 AM, ...
Calories                                   [81, 61, 59, 47, 48]

Id      Calories
count  2.209900e+04  22099.000000
mean   4.848235e+09     97.386760
std    2.422500e+09     60.702622
min    1.503960e+09     42.000000
25%    2.320127e+09     63.000000
50%    4.445115e+09     83.000000
75%    6.962181e+09    108.000000
max    8.877689e+09    948.000000

In [257]:
pstr(hourlyIntensities_merged)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22099 entries, 0 to 22098
Data columns (total 4 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Id                22099 non-null  int64  
 1   ActivityHour      22099 non-null  object 
 2   TotalIntensity    22099 non-null  int64  
 3   AverageIntensity  22099 non-null  float64
dtypes: float64(1), int64(2), object(1)
memory usage: 690.7+ KB
None


C:\Users\User\AppData\Local\Temp\ipykernel_10580\3310633354.py:4: FutureWarning: null_counts is deprecated. Use show_counts instead
  print(df.info(null_counts=True, verbose=True))


0
Id                [1503960366, 1624580081, 1644430081, 184450507...
ActivityHour      [4/12/2016 12:00:00 AM, 4/12/2016 1:00:00 AM, ...
TotalIntensity                                    [20, 8, 7, 0, 13]
AverageIntensity      [0.333333, 0.133333, 0.116667, 0.0, 0.216667]

Id  TotalIntensity  AverageIntensity
count  2.209900e+04    22099.000000      22099.000000
mean   4.848235e+09       12.035341          0.200589
std    2.422500e+09       21.133110          0.352219
min    1.503960e+09        0.000000          0.000000
25%    2.320127e+09        0.000000          0.000000
50%    4.445115e+09        3.000000          0.050000
75%    6.962181e+09       16.000000          0.266667
max    8.877689e+09      180.000000          3.000000

In [258]:
pstr(hourlySteps_merged)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22099 entries, 0 to 22098
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Id            22099 non-null  int64 
 1   ActivityHour  22099 non-null  object
 2   StepTotal     22099 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 518.1+ KB
None


C:\Users\User\AppData\Local\Temp\ipykernel_10580\3310633354.py:4: FutureWarning: null_counts is deprecated. Use show_counts instead
  print(df.info(null_counts=True, verbose=True))


0
Id            [1503960366, 1624580081, 1644430081, 184450507...
ActivityHour  [4/12/2016 12:00:00 AM, 4/12/2016 1:00:00 AM, ...
StepTotal                               [373, 160, 151, 0, 250]

Id     StepTotal
count  2.209900e+04  22099.000000
mean   4.848235e+09    320.166342
std    2.422500e+09    690.384228
min    1.503960e+09      0.000000
25%    2.320127e+09      0.000000
50%    4.445115e+09     40.000000
75%    6.962181e+09    357.000000
max    8.877689e+09  10554.000000

In [259]:
pstr(minuteCaloriesNarrow_merged)

C:\Users\User\AppData\Local\Temp\ipykernel_10580\3310633354.py:4: FutureWarning: null_counts is deprecated. Use show_counts instead
  print(df.info(null_counts=True, verbose=True))


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1325580 entries, 0 to 1325579
Data columns (total 3 columns):
 #   Column          Non-Null Count    Dtype  
---  ------          --------------    -----  
 0   Id              1325580 non-null  int64  
 1   ActivityMinute  1325580 non-null  object 
 2   Calories        1325580 non-null  float64
dtypes: float64(1), int64(1), object(1)
memory usage: 30.3+ MB
None


0
Id              [1503960366, 1624580081, 1644430081, 184450507...
ActivityMinute  [4/12/2016 12:00:00 AM, 4/12/2016 12:01:00 AM,...
Calories        [0.786499977111816, 0.94379997253418, 2.044899...

Id      Calories
count  1.325580e+06  1.325580e+06
mean   4.847898e+09  1.623130e+00
std    2.422313e+09  1.410447e+00
min    1.503960e+09  0.000000e+00
25%    2.320127e+09  9.357000e-01
50%    4.445115e+09  1.217600e+00
75%    6.962181e+09  1.432700e+00
max    8.877689e+09  1.974995e+01

In [260]:
pstr(minuteCaloriesWide_merged)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21645 entries, 0 to 21644
Data columns (total 62 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Id            21645 non-null  int64  
 1   ActivityHour  21645 non-null  object 
 2   Calories00    21645 non-null  float64
 3   Calories01    21645 non-null  float64
 4   Calories02    21645 non-null  float64
 5   Calories03    21645 non-null  float64
 6   Calories04    21645 non-null  float64
 7   Calories05    21645 non-null  float64
 8   Calories06    21645 non-null  float64
 9   Calories07    21645 non-null  float64
 10  Calories08    21645 non-null  float64
 11  Calories09    21645 non-null  float64
 12  Calories10    21645 non-null  float64
 13  Calories11    21645 non-null  float64
 14  Calories12    21645 non-null  float64
 15  Calories13    21645 non-null  float64
 16  Calories14    21645 non-null  float64
 17  Calories15    21645 non-null  float64
 18  Calories16    21645 non-nu

C:\Users\User\AppData\Local\Temp\ipykernel_10580\3310633354.py:4: FutureWarning: null_counts is deprecated. Use show_counts instead
  print(df.info(null_counts=True, verbose=True))


0
Id            [1503960366, 1624580081, 1644430081, 184450507...
ActivityHour  [4/13/2016 12:00:00 AM, 4/13/2016 1:00:00 AM, ...
Calories00    [1.88759994506836, 0.786499977111816, 0.943799...
Calories01    [2.20219993591309, 0.786499977111816, 2.044899...
Calories02    [0.94379997253418, 0.786499977111816, 2.044899...
...                                                         ...
Calories55    [1.88759994506836, 0.786499977111816, 0.943799...
Calories56    [0.94379997253418, 0.786499977111816, 2.516799...
Calories57    [0.94379997253418, 0.786499977111816, 2.044899...
Calories58    [0.94379997253418, 0.786499977111816, 2.831399...
Calories59    [0.94379997253418, 0.786499977111816, 2.202199...

[62 rows x 1 columns]

Id    Calories00    Calories01    Calories02    Calories03  \
count  2.164500e+04  21645.000000  21645.000000  21645.000000  21645.000000   
mean   4.836965e+09      1.622629      1.626377      1.637824      1.635515   
std    2.424088e+09      1.398418      1.395083      1.408828      1.419590   
min    1.503960e+09      0.702700      0.702700      0.702700      0.702700   
25%    2.320127e+09      0.935700      0.935700      0.937680      0.935700   
50%    4.445115e+09      1.217600      1.217600      1.220400      1.218500   
75%    6.962181e+09      1.432700      1.432700      1.432700      1.432700   
max    8.877689e+09     19.727337     19.727337     19.727337     19.727337   

         Calories04    Calories05    Calories06    Calories07    Calories08  \
count  21645.000000  21645.000000  21645.000000  21645.000000  21645.000000   
mean       1.637997      1.638306      1.639910      1.629520      1.623686   
std        1.433532      1.438253      1.435465      1.424092      1.411596   
min        0.702700      0.702700      0.702700      0.702700      0.702700   
25%        0.935700      0.935700      0.935700      0.935700      0.935700   
50%        1.218500      1.218500      1.218500      1.217600      1.217600   
75%        1.432700      1.432700      1.432700      1.432700      1.432700   
max       19.727337     19.727337     19.727337     19.727337     19.727337   

       ...    Calories50    Calories51    Calories52    Calories53  \
count  ...  21645.000000  21645.000000  21645.000000  21645.000000   
mean   ...      1.623665      1.613643      1.620958      1.618227   
std    ...      1.407171      1.395206      1.407914      1.400498   
min    ...      0.702700      0.702700      0.702700      0.702700   
25%    ...      0.935700      0.935700      0.935700      0.935700   
50%    ...      1.217600      1.217600      1.217600      1.217600   
75%    ...      1.432700      1.432700      1.432700      1.432700   
max    ...     19.749947     19.749947     19.749947     19.749947   

         Calories54    Calories55    Calories56    Calories57    Calories58  \
count  21645.000000  21645.000000  21645.000000  21645.000000  21645.000000   
mean       1.621229      1.615972      1.608714      1.612657      1.611715   
std        1.408974      1.392530      1.376827      1.369097      1.374954   
min        0.702700      0.702700      0.702700      0.702700      0.702700   
25%        0.935700      0.935700      0.935700      0.935700      0.935700   
50%        1.217600      1.217600      1.217600      1.217600      1.217600   
75%        1.432700      1.432700      1.432700      1.432700      1.432700   
max       19.749947     19.749947     19.727337     19.727337     19.727337   

         Calories59  
count  21645.000000  
mean       1.612110  
std        1.373888  
min        0.000000  
25%        0.935700  
50%        1.217600  
75%        1.432700  
max       19.727337  

[8 rows x 61 columns]

In [261]:
pstr(minuteIntensitiesNarrow_merged)

C:\Users\User\AppData\Local\Temp\ipykernel_10580\3310633354.py:4: FutureWarning: null_counts is deprecated. Use show_counts instead
  print(df.info(null_counts=True, verbose=True))


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1325580 entries, 0 to 1325579
Data columns (total 3 columns):
 #   Column          Non-Null Count    Dtype 
---  ------          --------------    ----- 
 0   Id              1325580 non-null  int64 
 1   ActivityMinute  1325580 non-null  object
 2   Intensity       1325580 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 30.3+ MB
None


0
Id              [1503960366, 1624580081, 1644430081, 184450507...
ActivityMinute  [4/12/2016 12:00:00 AM, 4/12/2016 12:01:00 AM,...
Intensity                                            [0, 1, 2, 3]

Id     Intensity
count  1.325580e+06  1.325580e+06
mean   4.847898e+09  2.005937e-01
std    2.422313e+09  5.190227e-01
min    1.503960e+09  0.000000e+00
25%    2.320127e+09  0.000000e+00
50%    4.445115e+09  0.000000e+00
75%    6.962181e+09  0.000000e+00
max    8.877689e+09  3.000000e+00

In [262]:
pstr(minuteIntensitiesWide_merged)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21645 entries, 0 to 21644
Data columns (total 62 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Id            21645 non-null  int64 
 1   ActivityHour  21645 non-null  object
 2   Intensity00   21645 non-null  int64 
 3   Intensity01   21645 non-null  int64 
 4   Intensity02   21645 non-null  int64 
 5   Intensity03   21645 non-null  int64 
 6   Intensity04   21645 non-null  int64 
 7   Intensity05   21645 non-null  int64 
 8   Intensity06   21645 non-null  int64 
 9   Intensity07   21645 non-null  int64 
 10  Intensity08   21645 non-null  int64 
 11  Intensity09   21645 non-null  int64 
 12  Intensity10   21645 non-null  int64 
 13  Intensity11   21645 non-null  int64 
 14  Intensity12   21645 non-null  int64 
 15  Intensity13   21645 non-null  int64 
 16  Intensity14   21645 non-null  int64 
 17  Intensity15   21645 non-null  int64 
 18  Intensity16   21645 non-null  int64 
 19  Inte

C:\Users\User\AppData\Local\Temp\ipykernel_10580\3310633354.py:4: FutureWarning: null_counts is deprecated. Use show_counts instead
  print(df.info(null_counts=True, verbose=True))


0
Id            [1503960366, 1624580081, 1644430081, 184450507...
ActivityHour  [4/13/2016 12:00:00 AM, 4/13/2016 1:00:00 AM, ...
Intensity00                                        [1, 0, 3, 2]
Intensity01                                        [1, 0, 3, 2]
Intensity02                                        [0, 1, 3, 2]
...                                                         ...
Intensity55                                        [1, 0, 2, 3]
Intensity56                                        [0, 1, 2, 3]
Intensity57                                        [0, 1, 3, 2]
Intensity58                                        [0, 1, 3, 2]
Intensity59                                        [0, 1, 3, 2]

[62 rows x 1 columns]

Id   Intensity00   Intensity01   Intensity02   Intensity03  \
count  2.164500e+04  21645.000000  21645.000000  21645.000000  21645.000000   
mean   4.836965e+09      0.199723      0.203326      0.208177      0.203835   
std    2.424088e+09      0.509819      0.515432      0.521394      0.518137   
min    1.503960e+09      0.000000      0.000000      0.000000      0.000000   
25%    2.320127e+09      0.000000      0.000000      0.000000      0.000000   
50%    4.445115e+09      0.000000      0.000000      0.000000      0.000000   
75%    6.962181e+09      0.000000      0.000000      0.000000      0.000000   
max    8.877689e+09      3.000000      3.000000      3.000000      3.000000   

        Intensity04   Intensity05   Intensity06   Intensity07   Intensity08  \
count  21645.000000  21645.000000  21645.000000  21645.000000  21645.000000   
mean       0.205082      0.204897      0.206560      0.201894      0.202310   
std        0.521956      0.521054      0.523053      0.519074      0.522594   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000      0.000000   
75%        0.000000      0.000000      0.000000      0.000000      0.000000   
max        3.000000      3.000000      3.000000      3.000000      3.000000   

       ...   Intensity50   Intensity51   Intensity52   Intensity53  \
count  ...  21645.000000  21645.000000  21645.000000  21645.000000   
mean   ...      0.201016      0.195796      0.198337      0.199399   
std    ...      0.514814      0.510299      0.511264      0.513331   
min    ...      0.000000      0.000000      0.000000      0.000000   
25%    ...      0.000000      0.000000      0.000000      0.000000   
50%    ...      0.000000      0.000000      0.000000      0.000000   
75%    ...      0.000000      0.000000      0.000000      0.000000   
max    ...      3.000000      3.000000      3.000000      3.000000   

        Intensity54   Intensity55   Intensity56   Intensity57   Intensity58  \
count  21645.000000  21645.000000  21645.000000  21645.000000  21645.000000   
mean       0.200139      0.198753      0.195565      0.199122      0.198244   
std        0.512142      0.511238      0.506435      0.511907      0.510124   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000      0.000000   
75%        0.000000      0.000000      0.000000      0.000000      0.000000   
max        3.000000      3.000000      3.000000      3.000000      3.000000   

        Intensity59  
count  21645.000000  
mean       0.195426  
std        0.503423  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max        3.000000  

[8 rows x 61 columns]

In [263]:
pstr(minuteMETsNarrow_merged)

C:\Users\User\AppData\Local\Temp\ipykernel_10580\3310633354.py:4: FutureWarning: null_counts is deprecated. Use show_counts instead
  print(df.info(null_counts=True, verbose=True))


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1325580 entries, 0 to 1325579
Data columns (total 3 columns):
 #   Column          Non-Null Count    Dtype 
---  ------          --------------    ----- 
 0   Id              1325580 non-null  int64 
 1   ActivityMinute  1325580 non-null  object
 2   METs            1325580 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 30.3+ MB
None


0
Id              [1503960366, 1624580081, 1644430081, 184450507...
ActivityMinute  [4/12/2016 12:00:00 AM, 4/12/2016 12:01:00 AM,...
METs                                         [10, 12, 26, 32, 36]

Id          METs
count  1.325580e+06  1.325580e+06
mean   4.847898e+09  1.469001e+01
std    2.422313e+09  1.205541e+01
min    1.503960e+09  0.000000e+00
25%    2.320127e+09  1.000000e+01
50%    4.445115e+09  1.000000e+01
75%    6.962181e+09  1.100000e+01
max    8.877689e+09  1.570000e+02

In [264]:
pstr(minuteSleep_merged)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 188521 entries, 0 to 188520
Data columns (total 4 columns):
 #   Column  Non-Null Count   Dtype 
---  ------  --------------   ----- 
 0   Id      188521 non-null  int64 
 1   date    188521 non-null  object
 2   value   188521 non-null  int64 
 3   logId   188521 non-null  int64 
dtypes: int64(3), object(1)
memory usage: 5.8+ MB
None


C:\Users\User\AppData\Local\Temp\ipykernel_10580\3310633354.py:4: FutureWarning: null_counts is deprecated. Use show_counts instead
  print(df.info(null_counts=True, verbose=True))


0
Id     [1503960366, 1644430081, 1844505072, 192797227...
date   [4/12/2016 2:47:30 AM, 4/12/2016 2:48:30 AM, 4...
value                                          [3, 2, 1]
logId  [11380564589, 11388770715, 11388770716, 114027...

Id          value         logId
count  1.885210e+05  188521.000000  1.885210e+05
mean   4.996595e+09       1.095793  1.149611e+10
std    2.066950e+09       0.328673  6.822863e+07
min    1.503960e+09       1.000000  1.137223e+10
25%    3.977334e+09       1.000000  1.143931e+10
50%    4.702922e+09       1.000000  1.150114e+10
75%    6.962181e+09       1.000000  1.155253e+10
max    8.792010e+09       3.000000  1.161625e+10

In [265]:
pstr(minuteStepsNarrow_merged)

C:\Users\User\AppData\Local\Temp\ipykernel_10580\3310633354.py:4: FutureWarning: null_counts is deprecated. Use show_counts instead
  print(df.info(null_counts=True, verbose=True))


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1325580 entries, 0 to 1325579
Data columns (total 3 columns):
 #   Column          Non-Null Count    Dtype 
---  ------          --------------    ----- 
 0   Id              1325580 non-null  int64 
 1   ActivityMinute  1325580 non-null  object
 2   Steps           1325580 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 30.3+ MB
None


0
Id              [1503960366, 1624580081, 1644430081, 184450507...
ActivityMinute  [4/12/2016 12:00:00 AM, 4/12/2016 12:01:00 AM,...
Steps                                          [0, 7, 32, 41, 29]

Id         Steps
count  1.325580e+06  1.325580e+06
mean   4.847898e+09  5.336192e+00
std    2.422313e+09  1.812830e+01
min    1.503960e+09  0.000000e+00
25%    2.320127e+09  0.000000e+00
50%    4.445115e+09  0.000000e+00
75%    6.962181e+09  0.000000e+00
max    8.877689e+09  2.200000e+02

In [266]:
pstr(minuteStepsWide_merged)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21645 entries, 0 to 21644
Data columns (total 62 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Id            21645 non-null  int64 
 1   ActivityHour  21645 non-null  object
 2   Steps00       21645 non-null  int64 
 3   Steps01       21645 non-null  int64 
 4   Steps02       21645 non-null  int64 
 5   Steps03       21645 non-null  int64 
 6   Steps04       21645 non-null  int64 
 7   Steps05       21645 non-null  int64 
 8   Steps06       21645 non-null  int64 
 9   Steps07       21645 non-null  int64 
 10  Steps08       21645 non-null  int64 
 11  Steps09       21645 non-null  int64 
 12  Steps10       21645 non-null  int64 
 13  Steps11       21645 non-null  int64 
 14  Steps12       21645 non-null  int64 
 15  Steps13       21645 non-null  int64 
 16  Steps14       21645 non-null  int64 
 17  Steps15       21645 non-null  int64 
 18  Steps16       21645 non-null  int64 
 19  Step

C:\Users\User\AppData\Local\Temp\ipykernel_10580\3310633354.py:4: FutureWarning: null_counts is deprecated. Use show_counts instead
  print(df.info(null_counts=True, verbose=True))


0
Id            [1503960366, 1624580081, 1644430081, 184450507...
ActivityHour  [4/13/2016 12:00:00 AM, 4/13/2016 1:00:00 AM, ...
Steps00                                       [4, 0, 37, 9, 91]
Steps01                                     [16, 0, 14, 11, 30]
Steps02                                      [0, 10, 30, 64, 6]
...                                                         ...
Steps55                                       [1, 0, 84, 7, 81]
Steps56                                      [0, 31, 5, 47, 27]
Steps57                                    [0, 11, 115, 14, 29]
Steps58                                   [0, 42, 105, 11, 130]
Steps59                                     [0, 16, 2, 105, 12]

[62 rows x 1 columns]

Id       Steps00       Steps01       Steps02       Steps03  \
count  2.164500e+04  21645.000000  21645.000000  21645.000000  21645.000000   
mean   4.836965e+09      5.304366      5.335412      5.531439      5.469439   
std    2.424088e+09     17.783331     17.678358     18.079791     18.106414   
min    1.503960e+09      0.000000      0.000000      0.000000      0.000000   
25%    2.320127e+09      0.000000      0.000000      0.000000      0.000000   
50%    4.445115e+09      0.000000      0.000000      0.000000      0.000000   
75%    6.962181e+09      0.000000      0.000000      0.000000      0.000000   
max    8.877689e+09    186.000000    180.000000    182.000000    182.000000   

            Steps04       Steps05       Steps06       Steps07      Steps08  \
count  21645.000000  21645.000000  21645.000000  21645.000000  21645.00000   
mean       5.461862      5.590252      5.559483      5.412474      5.35879   
std       18.288469     18.565165     18.484912     18.335665     18.20523   
min        0.000000      0.000000      0.000000      0.000000      0.00000   
25%        0.000000      0.000000      0.000000      0.000000      0.00000   
50%        0.000000      0.000000      0.000000      0.000000      0.00000   
75%        0.000000      0.000000      0.000000      0.000000      0.00000   
max      181.000000    180.000000    181.000000    183.000000    180.00000   

       ...       Steps50       Steps51       Steps52       Steps53  \
count  ...  21645.000000  21645.000000  21645.000000  21645.000000   
mean   ...      5.329175      5.194456      5.225595      5.145484   
std    ...     17.870527     17.601857     17.618497     17.570195   
min    ...      0.000000      0.000000      0.000000      0.000000   
25%    ...      0.000000      0.000000      0.000000      0.000000   
50%    ...      0.000000      0.000000      0.000000      0.000000   
75%    ...      0.000000      0.000000      0.000000      0.000000   
max    ...    182.000000    181.000000    181.000000    181.000000   

            Steps54       Steps55       Steps56       Steps57       Steps58  \
count  21645.000000  21645.000000  21645.000000  21645.000000  21645.000000   
mean       5.223654      5.281220      5.179533      5.251836      5.143636   
std       17.684634     17.828413     17.569268     17.686583     17.427494   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%        0.000000      0.000000      0.000000      0.000000      0.000000   
50%        0.000000      0.000000      0.000000      0.000000      0.000000   
75%        0.000000      0.000000      0.000000      0.000000      0.000000   
max      184.000000    181.000000    182.000000    182.000000    180.000000   

            Steps59  
count  21645.000000  
mean       5.288935  
std       17.721454  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max      189.000000  

[8 rows x 61 columns]

In [267]:
print(sleepDay_merged.info(null_counts=True, verbose=True))
display(sleepDay_merged.describe())

C:\Users\User\AppData\Local\Temp\ipykernel_10580\2059558041.py:1: FutureWarning: null_counts is deprecated. Use show_counts instead
  print(sleepDay_merged.info(null_counts=True, verbose=True))


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 413 entries, 0 to 412
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   Id                  413 non-null    int64 
 1   SleepDay            413 non-null    object
 2   TotalSleepRecords   413 non-null    int64 
 3   TotalMinutesAsleep  413 non-null    int64 
 4   TotalTimeInBed      413 non-null    int64 
dtypes: int64(4), object(1)
memory usage: 16.3+ KB
None


Id  TotalSleepRecords  TotalMinutesAsleep  TotalTimeInBed
count  4.130000e+02         413.000000          413.000000      413.000000
mean   5.000979e+09           1.118644          419.467312      458.639225
std    2.060360e+09           0.345521          118.344679      127.101607
min    1.503960e+09           1.000000           58.000000       61.000000
25%    3.977334e+09           1.000000          361.000000      403.000000
50%    4.702922e+09           1.000000          433.000000      463.000000
75%    6.962181e+09           1.000000          490.000000      526.000000
max    8.792010e+09           3.000000          796.000000      961.000000

In [268]:
pstr(weightLogInfo_merged)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 67 entries, 0 to 66
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Id              67 non-null     int64  
 1   Date            67 non-null     object 
 2   WeightKg        67 non-null     float64
 3   WeightPounds    67 non-null     float64
 4   Fat             2 non-null      float64
 5   BMI             67 non-null     float64
 6   IsManualReport  67 non-null     bool   
 7   LogId           67 non-null     int64  
dtypes: bool(1), float64(4), int64(2), object(1)
memory usage: 3.9+ KB
None


C:\Users\User\AppData\Local\Temp\ipykernel_10580\3310633354.py:4: FutureWarning: null_counts is deprecated. Use show_counts instead
  print(df.info(null_counts=True, verbose=True))


0
Id              [1503960366, 1927972279, 2873212765, 431970357...
Date            [5/2/2016 11:59:59 PM, 5/3/2016 11:59:59 PM, 4...
WeightKg        [52.5999984741211, 133.5, 56.7000007629395, 57...
WeightPounds    [115.963146545323, 294.317120016975, 125.00210...
Fat                                             [22.0, nan, 25.0]
BMI             [22.6499996185303, 47.5400009155273, 21.450000...
IsManualReport                                      [True, False]
LogId           [1462233599000, 1462319999000, 1460509732000, ...

Id    WeightKg  WeightPounds       Fat        BMI  \
count  6.700000e+01   67.000000     67.000000   2.00000  67.000000   
mean   7.009282e+09   72.035821    158.811801  23.50000  25.185224   
std    1.950322e+09   13.923206     30.695415   2.12132   3.066963   
min    1.503960e+09   52.599998    115.963147  22.00000  21.450001   
25%    6.962181e+09   61.400002    135.363832  22.75000  23.959999   
50%    6.962181e+09   62.500000    137.788914  23.50000  24.389999   
75%    8.877689e+09   85.049999    187.503152  24.25000  25.559999   
max    8.877689e+09  133.500000    294.317120  25.00000  47.540001   

              LogId  
count  6.700000e+01  
mean   1.461772e+12  
std    7.829948e+08  
min    1.460444e+12  
25%    1.461079e+12  
50%    1.461802e+12  
75%    1.462375e+12  
max    1.463098e+12

#### Clean the data

In [269]:
# Change the Id column to string
dailyActivity_merged['Id'] = dailyActivity_merged['Id'].astype(str)
dailyCalories_merged['Id'] = dailyCalories_merged['Id'].astype(str)
dailyIntensities_merged['Id'] = dailyIntensities_merged['Id'].astype(str)
dailySteps_merged['Id'] = dailySteps_merged['Id'].astype(str)
heartrate_seconds_merged['Id'] = heartrate_seconds_merged['Id'].astype(str)
hourlyCalories_merged['Id'] = hourlyCalories_merged['Id'].astype(str)
hourlyIntensities_merged['Id'] = hourlyIntensities_merged['Id'].astype(str)
hourlySteps_merged['Id'] = hourlySteps_merged['Id'].astype(str)
minuteCaloriesNarrow_merged['Id'] = minuteCaloriesNarrow_merged['Id'].astype(str)
minuteCaloriesWide_merged['Id'] = minuteCaloriesWide_merged['Id'].astype(str)
minuteIntensitiesNarrow_merged['Id'] = minuteIntensitiesNarrow_merged['Id'].astype(str)
minuteIntensitiesWide_merged['Id'] = minuteIntensitiesWide_merged['Id'].astype(str)
minuteMETsNarrow_merged['Id'] = minuteMETsNarrow_merged['Id'].astype(str)
minuteSleep_merged['Id'] = minuteSleep_merged['Id'].astype(str)
minuteStepsNarrow_merged['Id'] = minuteStepsNarrow_merged['Id'].astype(str)
minuteStepsWide_merged['Id'] = minuteStepsWide_merged['Id'].astype(str)
sleepDay_merged['Id'] = sleepDay_merged['Id'].astype(str)
weightLogInfo_merged['Id'] = weightLogInfo_merged['Id'].astype(str)


# Change the date and time format from string
format_second_p = '%m/%d/%Y %I:%M:%S %p'
format_mdy = '%m/%d/%Y'

dailyActivity_merged['ActivityDate'] = [datetime.strptime(x, format_mdy) for x in dailyActivity_merged['ActivityDate'].values]
dailyActivity_merged.rename(columns={'ActivityDate':'date'}, inplace=True)
dailyCalories_merged['ActivityDay'] = [datetime.strptime(x, format_mdy) for x in dailyCalories_merged['ActivityDay'].values]
dailyCalories_merged.rename(columns={'ActivityDay':'date'}, inplace=True)
dailyIntensities_merged['ActivityDay'] = [datetime.strptime(x, format_mdy) for x in dailyIntensities_merged['ActivityDay'].values]
dailyIntensities_merged.rename(columns={'ActivityDay':'date'}, inplace=True)
dailySteps_merged['ActivityDay'] = [datetime.strptime(x, format_mdy) for x in dailySteps_merged['ActivityDay'].values]
dailySteps_merged.rename(columns={'ActivityDay':'date'}, inplace=True)

hourlyCalories_merged['ActivityHour'] = [datetime.strptime(x, format_second_p) for x in hourlyCalories_merged['ActivityHour'].values]
hourlyCalories_merged.rename(columns={'ActivityHour':'datetime'}, inplace=True)
hourlyIntensities_merged['ActivityHour'] = [datetime.strptime(x, format_second_p) for x in hourlyIntensities_merged['ActivityHour'].values]
hourlyIntensities_merged.rename(columns={'ActivityHour':'datetime'}, inplace=True)
hourlySteps_merged['ActivityHour'] = [datetime.strptime(x, format_second_p) for x in hourlySteps_merged['ActivityHour'].values]
hourlySteps_merged.rename(columns={'ActivityHour':'datetime'}, inplace=True)

minuteCaloriesNarrow_merged['ActivityMinute'] = [datetime.strptime(x, format_second_p) for x in minuteCaloriesNarrow_merged['ActivityMinute'].values]
minuteCaloriesNarrow_merged.rename(columns={'ActivityMinute':'datetime'}, inplace=True)
minuteCaloriesWide_merged['ActivityHour'] = [datetime.strptime(x, format_second_p) for x in minuteCaloriesWide_merged['ActivityHour'].values]
minuteCaloriesWide_merged.rename(columns={'ActivityHour':'datetime'}, inplace=True)
minuteIntensitiesNarrow_merged['ActivityMinute'] = [datetime.strptime(x, format_second_p) for x in minuteIntensitiesNarrow_merged['ActivityMinute'].values]
minuteIntensitiesNarrow_merged.rename(columns={'ActivityMinute':'datetime'}, inplace=True)
minuteIntensitiesWide_merged['ActivityHour'] = [datetime.strptime(x, format_second_p) for x in minuteIntensitiesWide_merged['ActivityHour'].values]
minuteIntensitiesWide_merged.rename(columns={'ActivityHour':'datetime'}, inplace=True)
minuteStepsNarrow_merged['ActivityMinute'] = [datetime.strptime(x, format_second_p) for x in minuteStepsNarrow_merged['ActivityMinute'].values]
minuteStepsNarrow_merged.rename(columns={'ActivityMinute':'datetime'}, inplace=True)
minuteStepsWide_merged['ActivityHour'] = [datetime.strptime(x, format_second_p) for x in minuteStepsWide_merged['ActivityHour'].values]
minuteStepsWide_merged.rename(columns={'ActivityHour':'datetime'}, inplace=True)
sleepDay_merged['SleepDay'] = [datetime.strptime(x, format_second_p) for x in sleepDay_merged['SleepDay'].values]
sleepDay_merged.rename(columns={'SleepDay':'datetime'}, inplace=True)
weightLogInfo_merged['Date'] = [datetime.strptime(x, format_second_p) for x in weightLogInfo_merged['Date'].values]
weightLogInfo_merged.rename(columns={'Date':'datetime'}, inplace=True)


#### Daily Activities

##### dailyActivityMerged

In [271]:
dailyActivity_merged.head()

Id       date  TotalSteps  TotalDistance  TrackerDistance  \
0  1503960366 2016-04-12       13162           8.50             8.50   
1  1503960366 2016-04-13       10735           6.97             6.97   
2  1503960366 2016-04-14       10460           6.74             6.74   
3  1503960366 2016-04-15        9762           6.28             6.28   
4  1503960366 2016-04-16       12669           8.16             8.16   

   LoggedActivitiesDistance  VeryActiveDistance  ModeratelyActiveDistance  \
0                       0.0                1.88                      0.55   
1                       0.0                1.57                      0.69   
2                       0.0                2.44                      0.40   
3                       0.0                2.14                      1.26   
4                       0.0                2.71                      0.41   

   LightActiveDistance  SedentaryActiveDistance  VeryActiveMinutes  \
0                 6.06                      0.0                 25   
1                 4.71                      0.0                 21   
2                 3.91                      0.0                 30   
3                 2.83                      0.0                 29   
4                 5.04                      0.0                 36   

   FairlyActiveMinutes  LightlyActiveMinutes  SedentaryMinutes  Calories  
0                   13                   328               728      1985  
1                   19                   217               776      1797  
2                   11                   181              1218      1776  
3                   34                   209               726      1745  
4                   10                   221               773      1863

In [272]:
# dailyActivity_merged.groupby('Id').date.size().reset_index().sort_values(['date', 'Id'], ascending=[False, True])
x = dailyActivity_merged.groupby('Id').date.size().reset_index().groupby('date').Id.nunique().reset_index()
x['perc'] = np.round(x['Id']/x['Id'].sum()*100.0, 2)
x

date  Id   perc
0     4   1   3.03
1    18   1   3.03
2    19   1   3.03
3    20   1   3.03
4    26   2   6.06
5    28   1   3.03
6    29   2   6.06
7    30   3   9.09
8    31  21  63.64

In [273]:
dailyActivity_merged.select_dtypes('number')

TotalSteps  TotalDistance  TrackerDistance  LoggedActivitiesDistance  \
0         13162       8.500000         8.500000                       0.0   
1         10735       6.970000         6.970000                       0.0   
2         10460       6.740000         6.740000                       0.0   
3          9762       6.280000         6.280000                       0.0   
4         12669       8.160000         8.160000                       0.0   
..          ...            ...              ...                       ...   
935       10686       8.110000         8.110000                       0.0   
936       20226      18.250000        18.250000                       0.0   
937       10733       8.150000         8.150000                       0.0   
938       21420      19.559999        19.559999                       0.0   
939        8064       6.120000         6.120000                       0.0   

     VeryActiveDistance  ModeratelyActiveDistance  LightActiveDistance  \
0                  1.88                      0.55                 6.06   
1                  1.57                      0.69                 4.71   
2                  2.44                      0.40                 3.91   
3                  2.14                      1.26                 2.83   
4                  2.71                      0.41                 5.04   
..                  ...                       ...                  ...   
935                1.08                      0.20                 6.80   
936               11.10                      0.80                 6.24   
937                1.35                      0.46                 6.28   
938               13.22                      0.41                 5.89   
939                1.82                      0.04                 4.25   

     SedentaryActiveDistance  VeryActiveMinutes  FairlyActiveMinutes  \
0                       0.00                 25                   13   
1                       0.00                 21                   19   
2                       0.00                 30                   11   
3                       0.00                 29                   34   
4                       0.00                 36                   10   
..                       ...                ...                  ...   
935                     0.00                 17                    4   
936                     0.05                 73                   19   
937                     0.00                 18                   11   
938                     0.00                 88                   12   
939                     0.00                 23                    1   

     LightlyActiveMinutes  SedentaryMinutes  Calories  
0                     328               728      1985  
1                     217               776      1797  
2                     181              1218      1776  
3                     209               726      1745  
4                     221               773      1863  
..                    ...               ...       ...  
935                   245              1174      2847  
936                   217              1131      3710  
937                   224              1187      2832  
938                   213              1127      3832  
939                   137               770      1849  

[940 rows x 13 columns]

##### dailyCalories_merged

In [274]:
dailyCalories_merged.head()

Id       date  Calories
0  1503960366 2016-04-12      1985
1  1503960366 2016-04-13      1797
2  1503960366 2016-04-14      1776
3  1503960366 2016-04-15      1745
4  1503960366 2016-04-16      1863

##### dailyIntensities_merged

In [275]:
dailyIntensities_merged.head()

Id       date  SedentaryMinutes  LightlyActiveMinutes  \
0  1503960366 2016-04-12               728                   328   
1  1503960366 2016-04-13               776                   217   
2  1503960366 2016-04-14              1218                   181   
3  1503960366 2016-04-15               726                   209   
4  1503960366 2016-04-16               773                   221   

   FairlyActiveMinutes  VeryActiveMinutes  SedentaryActiveDistance  \
0                   13                 25                      0.0   
1                   19                 21                      0.0   
2                   11                 30                      0.0   
3                   34                 29                      0.0   
4                   10                 36                      0.0   

   LightActiveDistance  ModeratelyActiveDistance  VeryActiveDistance  
0                 6.06                      0.55                1.88  
1                 4.71                      0.69                1.57  
2                 3.91                      0.40                2.44  
3                 2.83                      1.26                2.14  
4                 5.04                      0.41                2.71

##### dailySteps_merged

In [276]:
dailySteps_merged.head()

Id       date  StepTotal
0  1503960366 2016-04-12      13162
1  1503960366 2016-04-13      10735
2  1503960366 2016-04-14      10460
3  1503960366 2016-04-15       9762
4  1503960366 2016-04-16      12669

In [277]:
dailyActivity_merged.shape, dailyCalories_merged.shape, dailyIntensities_merged.shape, dailySteps_merged.shape

((940, 15), (940, 3), (940, 10), (940, 3))

#### Hourly Activities

In [278]:
hourlyCalories_merged.head()

Id            datetime  Calories
0  1503960366 2016-04-12 00:00:00        81
1  1503960366 2016-04-12 01:00:00        61
2  1503960366 2016-04-12 02:00:00        59
3  1503960366 2016-04-12 03:00:00        47
4  1503960366 2016-04-12 04:00:00        48

In [279]:
hourlyIntensities_merged

Id            datetime  TotalIntensity  AverageIntensity
0      1503960366 2016-04-12 00:00:00              20          0.333333
1      1503960366 2016-04-12 01:00:00               8          0.133333
2      1503960366 2016-04-12 02:00:00               7          0.116667
3      1503960366 2016-04-12 03:00:00               0          0.000000
4      1503960366 2016-04-12 04:00:00               0          0.000000
...           ...                 ...             ...               ...
22094  8877689391 2016-05-12 10:00:00              12          0.200000
22095  8877689391 2016-05-12 11:00:00              29          0.483333
22096  8877689391 2016-05-12 12:00:00              93          1.550000
22097  8877689391 2016-05-12 13:00:00               6          0.100000
22098  8877689391 2016-05-12 14:00:00               9          0.150000

[22099 rows x 4 columns]

In [280]:
hourlySteps_merged.head()

Id            datetime  StepTotal
0  1503960366 2016-04-12 00:00:00        373
1  1503960366 2016-04-12 01:00:00        160
2  1503960366 2016-04-12 02:00:00        151
3  1503960366 2016-04-12 03:00:00          0
4  1503960366 2016-04-12 04:00:00          0

In [281]:
#How many rows in each file
hourlyCalories_merged.shape[0], hourlyIntensities_merged.shape[0], hourlySteps_merged.shape[0]

(22099, 22099, 22099)

In [282]:
#How many unique Ids in each file
hourlyCalories_merged.Id.drop_duplicates().shape[0], hourlyIntensities_merged.Id.drop_duplicates().shape[0], hourlySteps_merged.Id.drop_duplicates().shape[0]

(33, 33, 33)

In [283]:
#Date Range in each file
hourlyCalories_merged.datetime.min(), hourlyCalories_merged.datetime.max(), \
    hourlyIntensities_merged.datetime.min(), hourlyIntensities_merged.datetime.max(), \
        hourlySteps_merged.datetime.min(), hourlySteps_merged.datetime.max()

(Timestamp('2016-04-12 00:00:00'),
 Timestamp('2016-05-12 15:00:00'),
 Timestamp('2016-04-12 00:00:00'),
 Timestamp('2016-05-12 15:00:00'),
 Timestamp('2016-04-12 00:00:00'),
 Timestamp('2016-05-12 15:00:00'))

#### Minute Activities

In [284]:
# Sort by Ids and time
minuteCaloriesNarrow_merged = minuteCaloriesNarrow_merged.sort_values(['Id', 'datetime'])
minuteCaloriesWide_merged = minuteCaloriesWide_merged.sort_values(['Id', 'datetime'])
minuteIntensitiesNarrow_merged = minuteIntensitiesNarrow_merged.sort_values(['Id', 'datetime'])
minuteIntensitiesWide_merged = minuteIntensitiesWide_merged.sort_values(['Id', 'datetime'])
minuteStepsNarrow_merged = minuteStepsNarrow_merged.sort_values(['Id', 'datetime'])
minuteStepsWide_merged = minuteStepsWide_merged.sort_values(['Id', 'datetime'])

In [285]:
minuteCaloriesNarrow_merged.head()

Id            datetime  Calories
0  1503960366 2016-04-12 00:00:00    0.7865
1  1503960366 2016-04-12 00:01:00    0.7865
2  1503960366 2016-04-12 00:02:00    0.7865
3  1503960366 2016-04-12 00:03:00    0.7865
4  1503960366 2016-04-12 00:04:00    0.7865

In [286]:
minuteCaloriesWide_merged.head()

Id            datetime  Calories00  Calories01  Calories02  \
0  1503960366 2016-04-13 00:00:00      1.8876      2.2022      0.9438   
1  1503960366 2016-04-13 01:00:00      0.7865      0.7865      0.7865   
2  1503960366 2016-04-13 02:00:00      0.7865      0.7865      0.7865   
3  1503960366 2016-04-13 03:00:00      0.7865      0.7865      0.7865   
4  1503960366 2016-04-13 04:00:00      0.7865      0.7865      0.7865   

   Calories03  Calories04  Calories05  Calories06  Calories07  ...  \
0      0.9438      0.9438      2.0449      0.9438      2.2022  ...   
1      0.7865      0.9438      0.9438      0.9438      0.7865  ...   
2      0.7865      0.7865      0.7865      0.7865      0.7865  ...   
3      0.7865      0.7865      0.7865      0.7865      0.7865  ...   
4      0.7865      0.7865      0.7865      0.7865      0.7865  ...   

   Calories50  Calories51  Calories52  Calories53  Calories54  Calories55  \
0      0.9438      2.0449      2.0449      0.9438      2.3595      1.8876   
1      0.7865      0.7865      0.7865      0.7865      0.7865      0.7865   
2      0.7865      0.7865      0.7865      0.7865      0.7865      0.7865   
3      0.7865      0.7865      0.7865      0.7865      0.7865      0.7865   
4      0.7865      0.7865      0.7865      0.7865      0.7865      0.7865   

   Calories56  Calories57  Calories58  Calories59  
0      0.9438      0.9438      0.9438      0.9438  
1      0.7865      0.7865      0.7865      0.7865  
2      0.7865      0.7865      0.7865      0.7865  
3      0.7865      0.7865      0.7865      0.7865  
4      0.7865      0.7865      0.7865      0.7865  

[5 rows x 62 columns]

In [287]:
minuteIntensitiesNarrow_merged.head()

Id            datetime  Intensity
0  1503960366 2016-04-12 00:00:00          0
1  1503960366 2016-04-12 00:01:00          0
2  1503960366 2016-04-12 00:02:00          0
3  1503960366 2016-04-12 00:03:00          0
4  1503960366 2016-04-12 00:04:00          0

In [288]:
minuteIntensitiesWide_merged.head()

Id            datetime  Intensity00  Intensity01  Intensity02  \
0  1503960366 2016-04-13 00:00:00            1            1            0   
1  1503960366 2016-04-13 01:00:00            0            0            0   
2  1503960366 2016-04-13 02:00:00            0            0            0   
3  1503960366 2016-04-13 03:00:00            0            0            0   
4  1503960366 2016-04-13 04:00:00            0            0            0   

   Intensity03  Intensity04  Intensity05  Intensity06  Intensity07  ...  \
0            0            0            1            0            1  ...   
1            0            0            0            0            0  ...   
2            0            0            0            0            0  ...   
3            0            0            0            0            0  ...   
4            0            0            0            0            0  ...   

   Intensity50  Intensity51  Intensity52  Intensity53  Intensity54  \
0            0            1            1            0            1   
1            0            0            0            0            0   
2            0            0            0            0            0   
3            0            0            0            0            0   
4            0            0            0            0            0   

   Intensity55  Intensity56  Intensity57  Intensity58  Intensity59  
0            1            0            0            0            0  
1            0            0            0            0            0  
2            0            0            0            0            0  
3            0            0            0            0            0  
4            0            0            0            0            0  

[5 rows x 62 columns]

In [289]:
minuteStepsNarrow_merged.head()

Id            datetime  Steps
0  1503960366 2016-04-12 00:00:00      0
1  1503960366 2016-04-12 00:01:00      0
2  1503960366 2016-04-12 00:02:00      0
3  1503960366 2016-04-12 00:03:00      0
4  1503960366 2016-04-12 00:04:00      0

In [290]:
minuteStepsWide_merged.head()

Id            datetime  Steps00  Steps01  Steps02  Steps03  \
0  1503960366 2016-04-13 00:00:00        4       16        0        0   
1  1503960366 2016-04-13 01:00:00        0        0        0        0   
2  1503960366 2016-04-13 02:00:00        0        0        0        0   
3  1503960366 2016-04-13 03:00:00        0        0        0        0   
4  1503960366 2016-04-13 04:00:00        0        0        0        0   

   Steps04  Steps05  Steps06  Steps07  ...  Steps50  Steps51  Steps52  \
0        0        9        0       17  ...        0        9        8   
1        0        0        0        0  ...        0        0        0   
2        0        0        0        0  ...        0        0        0   
3        0        0        0        0  ...        0        0        0   
4        0        0        0        0  ...        0        0        0   

   Steps53  Steps54  Steps55  Steps56  Steps57  Steps58  Steps59  
0        0       20        1        0        0        0        0  
1        0        0        0        0        0        0        0  
2        0        0        0        0        0        0        0  
3        0        0        0        0        0        0        0  
4        0        0        0        0        0        0        0  

[5 rows x 62 columns]

In [291]:
minuteMETsNarrow_merged.head()

Id         ActivityMinute  METs
0  1503960366  4/12/2016 12:00:00 AM    10
1  1503960366  4/12/2016 12:01:00 AM    10
2  1503960366  4/12/2016 12:02:00 AM    10
3  1503960366  4/12/2016 12:03:00 AM    10
4  1503960366  4/12/2016 12:04:00 AM    10

In [292]:
minuteSleep_merged.head()

Id                  date  value        logId
0  1503960366  4/12/2016 2:47:30 AM      3  11380564589
1  1503960366  4/12/2016 2:48:30 AM      2  11380564589
2  1503960366  4/12/2016 2:49:30 AM      1  11380564589
3  1503960366  4/12/2016 2:50:30 AM      1  11380564589
4  1503960366  4/12/2016 2:51:30 AM      1  11380564589

In [293]:
#How many rows in each file
minuteCaloriesNarrow_merged.shape, minuteCaloriesWide_merged.shape, \
    minuteIntensitiesNarrow_merged.shape, minuteIntensitiesWide_merged.shape, \
        minuteStepsNarrow_merged.shape, minuteStepsWide_merged.shape

((1325580, 3),
 (21645, 62),
 (1325580, 3),
 (21645, 62),
 (1325580, 3),
 (21645, 62))

In [294]:
#How many unique Ids in each files
minuteCaloriesNarrow_merged.Id.drop_duplicates().shape[0], minuteCaloriesWide_merged.Id.drop_duplicates().shape[0], \
    minuteIntensitiesNarrow_merged.Id.drop_duplicates().shape[0], minuteIntensitiesWide_merged.Id.drop_duplicates().shape[0], \
        minuteStepsNarrow_merged.Id.drop_duplicates().shape[0], minuteStepsWide_merged.Id.drop_duplicates().shape[0]

(33, 33, 33, 33, 33, 33)

In [295]:
#Date Range of each files
minuteCaloriesNarrow_merged.datetime.min(), minuteCaloriesNarrow_merged.datetime.max(), \
    minuteCaloriesWide_merged.datetime.min(), minuteCaloriesWide_merged.datetime.max(), \
        minuteIntensitiesNarrow_merged.datetime.min(), minuteIntensitiesNarrow_merged.datetime.max(), \
            minuteIntensitiesWide_merged.datetime.min(), minuteIntensitiesWide_merged.datetime.max(), \
        minuteStepsNarrow_merged.datetime.min(), minuteStepsNarrow_merged.datetime.max(), \
            minuteStepsWide_merged.datetime.min(), minuteStepsWide_merged.datetime.max()

(Timestamp('2016-04-12 00:00:00'),
 Timestamp('2016-05-12 15:59:00'),
 Timestamp('2016-04-13 00:00:00'),
 Timestamp('2016-05-13 08:00:00'),
 Timestamp('2016-04-12 00:00:00'),
 Timestamp('2016-05-12 15:59:00'),
 Timestamp('2016-04-13 00:00:00'),
 Timestamp('2016-05-13 08:00:00'),
 Timestamp('2016-04-12 00:00:00'),
 Timestamp('2016-05-12 15:59:00'),
 Timestamp('2016-04-13 00:00:00'),
 Timestamp('2016-05-13 08:00:00'))

#### Others

In [328]:
# Add Hours columns to sleepDay_merged to make analysis much easier
sleepDay_merged['TotalHoursAsleep'] = sleepDay_merged['TotalMinutesAsleep']/60.0
sleepDay_merged['TotalHoursInBed'] = sleepDay_merged['TotalTimeInBed']/60.

#Add difference of Hours Asleep with Hours in Bed
sleepDay_merged['sleep_bed_diff'] = sleepDay_merged['TotalHoursInBed'] - sleepDay_merged['TotalHoursAsleep']

#Add day of weeks into it
sleepDay_merged['weekday'] = sleepDay_merged['datetime'].dt.day_name()

conditions = [(sleepDay_merged['weekday'] == 'Sunday'),
            (sleepDay_merged['weekday'] == 'Monday'),
            (sleepDay_merged['weekday'] == 'Tuesday'),
            (sleepDay_merged['weekday'] == 'Wednesday'),
            (sleepDay_merged['weekday'] == 'Thursday'),
            (sleepDay_merged['weekday'] == 'Friday'),
            (sleepDay_merged['weekday'] == 'Saturday')]

choices = [7, 1, 2, 3, 4, 5, 6]

sleepDay_merged['week'] = np.select(conditions, choices)

sleepDay_merged.head()

Id   datetime  TotalSleepRecords  TotalMinutesAsleep  \
0  1503960366 2016-04-12                  1                 327   
1  1503960366 2016-04-13                  2                 384   
2  1503960366 2016-04-15                  1                 412   
3  1503960366 2016-04-16                  2                 340   
4  1503960366 2016-04-17                  1                 700   

   TotalTimeInBed  TotalHoursAsleep  TotalHoursInBed  sleep_bed_diff  \
0             346          5.450000         5.766667        0.316667   
1             407          6.400000         6.783333        0.383333   
2             442          6.866667         7.366667        0.500000   
3             367          5.666667         6.116667        0.450000   
4             712         11.666667        11.866667        0.200000   

     weekday  week  
0    Tuesday     2  
1  Wednesday     3  
2     Friday     5  
3   Saturday     6  
4     Sunday     7

In [307]:
x = sleepDay_merged.groupby('Id').datetime.nunique().reset_index().groupby('datetime').Id.nunique().reset_index()
x['perc'] = np.round(x['Id']/x['Id'].sum()*100)
print(x.Id.sum())
display(x)

24


datetime  Id  perc
0          1   1   4.0
1          2   1   4.0
2          3   3  12.0
3          4   1   4.0
4          5   2   8.0
5          8   1   4.0
6         15   2   8.0
7         18   1   4.0
8         23   1   4.0
9         24   1   4.0
10        25   1   4.0
11        26   2   8.0
12        27   1   4.0
13        28   3  12.0
14        31   3  12.0

Only 24 users has ther sleep data recorded. And 24% of those users (6 Ids) recorded at least 28 days of data. 

In [315]:
x = sleepDay_merged.groupby('Id').TotalSleepRecords.mean().reset_index().groupby('TotalSleepRecords').Id.nunique().reset_index()
x['perc'] = np.round(x['Id']/x['Id'].sum()*100)
display(x)
x = sleepDay_merged.groupby('Id').agg({'TotalSleepRecords': np.mean, 'datetime': 'size'}).reset_index().sort_values('TotalSleepRecords', ascending=False).head(5)
x.rename(columns={'TotalSleepRecords':'Avg_TotalSleepRecords', 'datetime':'ndays'}, inplace=True)
display(x)

TotalSleepRecords  Id  perc
0            1.000000  12  50.0
1            1.038462   2   8.0
2            1.071429   1   4.0
3            1.080000   1   4.0
4            1.096774   1   4.0
5            1.125000   1   4.0
6            1.142857   1   4.0
7            1.222222   1   4.0
8            1.225806   1   4.0
9            1.291667   1   4.0
10           1.392857   1   4.0
11           1.600000   1   4.0

Id  Avg_TotalSleepRecords  ndays
3   1927972279               1.600000      5
11  4445114986               1.392857     28
10  4388161847               1.291667     24
14  5553957443               1.225806     31
16  6117666160               1.222222     18

About 12% of us the sample users sleeps only once during the sample timespan. The other 50% has slept more than once in at least one day during this time period.

In [330]:
display(sleepDay_merged.agg({'TotalHoursAsleep': [np.min, np.mean, np.median, np.max], \
    'TotalHoursInBed': [np.min, np.mean, np.median, np.max], \
    'sleep_bed_diff': [np.min, np.mean, np.median, np.max]}).reset_index())
display(sleepDay_merged.groupby('Id').agg({'TotalHoursAsleep': ['size', np.min, np.mean, np.median, np.max], \
    'TotalHoursInBed': [np.min, np.mean, np.median, np.max], \
    'sleep_bed_diff': [np.min, np.mean, np.median, np.max]}).reset_index())
x = sleepDay_merged.query('TotalHoursAsleep < 7')
print("{} users have slept less than 7 hours at least once in the time spand".format(x.Id.nunique()))
x = sleepDay_merged.query('TotalHoursAsleep < 3')
print("{} users have slept less than 3 hours at least once in the time spand".format(x.Id.nunique()))
x = sleepDay_merged.groupby('Id').TotalHoursAsleep.mean().reset_index().query('TotalHoursAsleep < 7.0')
print("{} users have slept on average less than 7 hours in the time spand".format(x.Id.nunique()))
x = sleepDay_merged.groupby('Id').TotalHoursAsleep.mean().reset_index().query('TotalHoursAsleep < 3.0')
print("{} users have slept on average less than 3 hours in the time spand".format(x.Id.nunique()))

index  TotalHoursAsleep  TotalHoursInBed  sleep_bed_diff
0    amin          0.966667         1.016667        0.000000
1    mean          6.991122         7.643987        0.652865
2  median          7.216667         7.716667        0.416667
3    amax         13.266667        16.016667        6.183333

Id TotalHoursAsleep                                             \
                           size      amin       mean     median       amax   
0   1503960366               25  4.083333   6.004667   5.666667  11.666667   
1   1644430081                4  1.983333   4.900000   2.175000  13.266667   
2   1844505072                3  9.833333  10.866667  10.733333  12.033333   
3   1927972279                5  2.766667   6.950000   6.633333  12.500000   
4   2026352035               28  5.950000   8.436310   8.608333   9.550000   
5   2320127002                1  1.016667   1.016667   1.016667   1.016667   
6   2347167796               15  6.233333   7.446667   7.416667   9.266667   
7   3977333714               28  2.533333   4.894048   4.858333   7.066667   
8   4020332650                8  1.283333   5.822917   6.241667   8.350000   
9   4319703577               26  0.983333   7.944231   8.225000  11.533333   
10  4388161847               24  1.033333   6.718750   7.225000  10.316667   
11  4445114986               28  1.633333   6.419643   6.716667   8.366667   
12  4558609924                5  1.716667   2.126667   2.050000   2.850000   
13  4702921684               28  4.216667   7.019048   7.050000   9.850000   
14  5553957443               31  5.366667   7.724731   7.300000  12.916667   
15  5577150313               26  1.233333   7.200000   7.125000  10.050000   
16  6117666160               18  5.600000   7.979630   8.200000  10.966667   
17  6775888955                3  3.916667   5.827778   6.516667   7.050000   
18  6962181067               31  4.966667   7.466667   7.366667  10.500000   
19  7007744171                2  0.966667   1.141667   1.141667   1.316667   
20  7086361926               24  5.366667   7.552083   7.516667  11.350000   
21  8053475328                3  1.233333   4.950000   5.516667   8.100000   
22  8378563200               32  5.383333   7.389062   7.350000  10.183333   
23  8792009665               15  5.650000   7.261111   7.050000   8.850000   

   TotalHoursInBed                                  sleep_bed_diff            \
              amin       mean     median       amax           amin      mean   
0         4.400000   6.386667   6.116667  11.866667       0.066667  0.382000   
1         2.116667   5.766667   2.466667  16.016667       0.133333  0.866667   
2        16.016667  16.016667  16.016667  16.016667       3.983333  5.150000   
3         2.966667   7.296667   7.033333  12.916667       0.200000  0.346667   
4         6.333333   8.960714   9.091667  10.116667       0.166667  0.524405   
5         1.150000   1.150000   1.150000   1.150000       0.133333  0.133333   
6         6.433333   8.188889   8.150000  10.033333       0.200000  0.742222   
7         5.083333   7.685714   7.800000  10.433333       1.766667  2.791667   
8         1.283333   6.329167   6.750000   9.016667       0.000000  0.506250   
9         1.083333   8.366026   8.608333  12.033333       0.100000  0.421795   
10        1.083333   7.103472   7.633333  10.683333       0.050000  0.384722   
11        1.783333   6.947024   7.183333   9.033333       0.033333  0.527381   
12        2.016667   2.333333   2.233333   2.983333       0.133333  0.206667   
13        4.283333   7.366071   7.341667  10.200000       0.066667  0.347024   
14        5.883333   8.431183   7.833333  14.050000       0.333333  0.706452   
15        1.300000   7.676923   7.658333  10.566667       0.066667  0.476923   
16        5.833333   8.502778   8.600000  11.633333       0.150000  0.523148   
17        4.333333   6.150000   6.766667   7.350000       0.250000  0.322222   
18        5.566667   7.768817   7.616667  11.316667       0.033333  0.302151   
19        1.016667   1.191667   1.191667   1.366667       0.050000  0.050000   
20        5.550000   7.773611   7.700000  11.733333       0.016667  0.221528   
21        1.250000   5.027778   5.616667   8.216667       0.016667  0.077778   
22        5.916667   8.055208   7.883333  11.483333       0.233

23 users have slept less than 7 hours at least once in the time spand
12 users have slept less than 3 hours at least once in the time spand
12 users have slept on average less than 7 hours in the time spand
3 users have slept on average less than 3 hours in the time spand


We can see the amount of hours each users have spend each day as well as the hours in bed. Almost all users (23) have slept less than 7 hours. And almost half (12 users) have slept only less than 3 hours. On average across the entire time span of daa collection, there are 12 users (50%) who slept less than 7 hours and 3 of them only slept 3 hours on average. 

##### IMPORTANT: **This analysis shows that users are having trouble getting the adequate amount of sleep**

We can also see the difference of hours spent on bed vs actual sleep. On average, users spent *half an hour* more in bed. There also seems to be an anomaly in one of the users time spent in bed vs sleep. User `3977333714` and `1844505072` spends an additional 2.7 hours and 5 hours, respectively, in bed on average ! This could be an indicator of bad sleep habits, case of insomnia, or just bad data record. Either way, it warants a deeper dive into these users' data and device.

In [297]:
weightLogInfo_merged.head(20)

Id            datetime    WeightKg  WeightPounds   Fat        BMI  \
0   1503960366 2016-05-02 23:59:59   52.599998    115.963147  22.0  22.650000   
1   1503960366 2016-05-03 23:59:59   52.599998    115.963147   NaN  22.650000   
2   1927972279 2016-04-13 01:08:52  133.500000    294.317120   NaN  47.540001   
3   2873212765 2016-04-21 23:59:59   56.700001    125.002104   NaN  21.450001   
4   2873212765 2016-05-12 23:59:59   57.299999    126.324875   NaN  21.690001   
5   4319703577 2016-04-17 23:59:59   72.400002    159.614681  25.0  27.450001   
6   4319703577 2016-05-04 23:59:59   72.300003    159.394222   NaN  27.379999   
7   4558609924 2016-04-18 23:59:59   69.699997    153.662190   NaN  27.250000   
8   4558609924 2016-04-25 23:59:59   70.300003    154.984977   NaN  27.459999   
9   4558609924 2016-05-01 23:59:59   69.900002    154.103125   NaN  27.320000   
10  4558609924 2016-05-02 23:59:59   69.199997    152.559879   NaN  27.040001   
11  4558609924 2016-05-09 23:59:59   69.099998    152.339420   NaN  27.000000   
12  5577150313 2016-04-17 09:17:55   90.699997    199.959265   NaN  28.000000   
13  6962181067 2016-04-12 23:59:59   62.500000    137.788914   NaN  24.389999   
14  6962181067 2016-04-13 23:59:59   62.099998    136.907061   NaN  24.240000   
15  6962181067 2016-04-14 23:59:59   61.700001    136.025217   NaN  24.100000   
16  6962181067 2016-04-15 23:59:59   61.500000    135.584291   NaN  24.000000   
17  6962181067 2016-04-16 23:59:59   62.000000    136.686603   NaN  24.209999   
18  6962181067 2016-04-17 23:59:59   61.400002    135.363832   NaN  23.959999   
19  6962181067 2016-04-18 23:59:59   61.200001    134.922906   NaN  23.889999   

    IsManualReport          LogId  
0             True  1462233599000  
1             True  1462319999000  
2            False  1460509732000  
3             True  1461283199000  
4             True  1463097599000  
5             True  1460937599000  
6             True  1462406399000  
7             True  1461023999000  
8             True  1461628799000  
9             True  1462147199000  
10            True  1462233599000  
11            True  1462838399000  
12           False  1460884675000  
13            True  1460505599000  
14            True  1460591999000  
15            True  1460678399000  
16            True  1460764799000  
17            True  1460851199000  
18            True  1460937599000  
19            True  1461023999000

In [298]:
sleepDay_merged.shape, weightLogInfo_merged.shape

((413, 5), (67, 8))

In [299]:
sleepDay_merged.Id.drop_duplicates().shape, weightLogInfo_merged.Id.drop_duplicates().shape

((24,), (8,))